listenbrainz/listenstore/timescale_listenstore.py

```diff
@@ -208,94 +246,158 @@ def insert(self, listens):
-            elif from_ts is not None:
-                query += "AND listened_at > :from_ts "
-            else:
-                query += "AND listened_at < :to_ts "
+                    WHERE user_name IN :user_names
+                      AND listened_at > :from_ts
+                      AND listened_at < :to_ts
+                 ORDER BY listened_at """ + ORDER_TEXT[order] + " LIMIT :limit"
+
+        if from_ts and to_ts:
+            to_dynamic = False
+            from_dynamic = False
+        elif from_ts is not None:
+            to_ts = from_ts + window_size
+            to_dynamic = True
+            from_dynamic = False
         else:
-            query += """AND listened_at > :from_ts
-                        AND listened_at < :to_ts """
-
-        query += "ORDER BY listened_at " + ORDER_TEXT[order] + " LIMIT :limit"
+            from_ts = to_ts - window_size
+            to_dynamic = False
+            from_dynamic = True
 
         listens = []
+        done = False
         with timescale.engine.connect() as connection:
-            curs = connection.execute(sqlalchemy.text(query), user_names=tuple(user_names), from_ts=from_ts, to_ts=to_ts, limit=limit)
+            t0 = time.time()
+
+            passes = 0
             while True:
-                result = curs.fetchone()
-                if not result:
+                passes += 1
+
+                # Oh shit valve. I'm keeping it here for the time being. :)
+                if passes == 10:
+                    done = True
                     break
 
-                listens.append(Listen.from_timescale(result[0], result[1], result[4], result[2], result[3]))
+                curs = connection.execute(sqlalchemy.text(query), user_names=tuple(user_names),
+                                          from_ts=from_ts, to_ts=to_ts, limit=limit)
+                while True:
+                    result = curs.fetchone()
+                    if not result:
+                        if not to_dynamic and not from_dynamic:
+                            done = True
+                            break
+
+                        if from_ts < min_user_ts - 1:
+                            done = True
+                            break
+
+                        if to_ts > int(time.time()) + ListenStore.MAX_FUTURE_SECONDS:
+                            done = True
+                            break
+
+                        if to_dynamic:
+                            from_ts += window_size - 1
+                            window_size *= WINDOW_SIZE_MULTIPLIER
+                            to_ts += window_size
+
+                        if from_dynamic:
+                            to_ts -= window_size
+                            window_size *= WINDOW_SIZE_MULTIPLIER
+                            from_ts -= window_size
+
+                        break
+
+                    listens.append(Listen.from_timescale(
+                        result[0], result[1], result[4], result[2], result[3]))
```

Comment:
I see that this is copied from an existing line, but seeing rows added in a different order to what they were selected from seems like a recipe for disaster. We should add it to an issue to fix at a later stage

Reply:
Reordered.